x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java

```diff
@@ -266,12 +277,45 @@ public void onIndexModule(IndexModule indexModule) {
         });
     }
 
+    static final Comparator<String> SEGMENT_FILENAME_COMPARATOR = Comparator.comparingLong(SegmentInfos::generationFromSegmentsFileName);
+
     @Override
     public Optional<EngineFactory> getEngineFactory(IndexSettings indexSettings) {
         if (SearchableSnapshotsConstants.isSearchableSnapshotStore(indexSettings.getSettings())
             && indexSettings.getSettings().getAsBoolean("index.frozen", false) == false) {
             return Optional.of(
-                engineConfig -> new ReadOnlyEngine(engineConfig, null, new TranslogStats(), false, Function.identity(), false)
+                engineConfig -> new ReadOnlyEngine(engineConfig, null, new TranslogStats(), false, Function.identity(), false) {
+
+                    // present an empty IndexCommit to the snapshot mechanism so that we copy no shard data to the repository
+                    private final IndexCommit emptyIndexCommit;
+
+                    {
+                        try {
+                            // We have to use a generation number that corresponds to a real segments_N file since we read this file from
+                            // the directory during the snapshotting process. The oldest segments_N file is the one in the snapshot
+                            // (recalling that we may perform some no-op commits which make newer segments_N files too). The good thing
+                            // about using the oldest segments_N file is that a restore will find that we already have this file "locally",
+                            // avoid overwriting the real one with the bogus one, and then use the real one for the rest of the recovery.
+
+                            final Directory directory = engineConfig.getStore().directory();
+                            final String oldestSegmentsFile = Arrays.stream(directory.listAll())
+                                .filter(s -> s.startsWith(IndexFileNames.SEGMENTS + "_"))
+                                .min(SEGMENT_FILENAME_COMPARATOR)
+                                .orElseThrow(() -> new IOException("segments_N file not found"));
+                            final SegmentInfos segmentInfos = new SegmentInfos(Version.LATEST.major);
+                            segmentInfos.updateGeneration(SegmentInfos.readCommit(directory, oldestSegmentsFile));
+                            emptyIndexCommit = Lucene.getIndexCommit(segmentInfos, directory);
+                        } catch (IOException e) {
+                            throw new UncheckedIOException(e);
+                        }
+                    }
+
+                    @Override
+                    public IndexCommitRef acquireIndexCommitForSnapshot() throws EngineException {
+                        store.incRef();
+                        return new IndexCommitRef(emptyIndexCommit, store::decRef);
```

Comment:
If we can skip the restore completely (see previous comment), could we then instead return null here and also skip the snapshotting completely?

Reply:
Not trivially, no, we want the snapshot of this shard to succeed (without having done much) so we can't just bail out, and we use a `null` commit to indicate "find the latest commit" further down the line.